llm_bench/python/utils/ov_utils.py

```diff
@@ -153,83 +251,98 @@ def greedy_search(input_port):
-        beam_idx_exist = 'beam_idx' in [port.any_name for port in ov_model.inputs]
-        assert num_beams == 1 or beam_idx_exist, 'Requested to make_stateful with num_beams > 1 but there is no beam_idx parameter for cache reorder fused'
-        # Set batch size for input_ids and attention mask to avoid dynamic dimension got propagated from the end of the model back to ReadValue
-        for input in not_kv_inputs:
-            shape = input.get_partial_shape()
-            if shape.rank.get_length() <= 2:  # == 1 for beam_index
-                shape[0] = batch_size * num_beams
-                input.get_node().set_partial_shape(shape)
-            else:
-                print(f'[ WARNING ] Rank of {input.get_any_name()} input of the model is not 2, batch size is not set')
+    # By default, batch is the 0-th but chatglm uses 1-st dimension as batch
+    # TODO: Deduce from a model via ordinal reshape (?) and topology
+    batch_dim = 1 if hf_model.config.model_type == 'chatglm' else 0
 
-        for kv_name_pair in zip(hf_model.key_value_input_names, hf_model.key_value_output_names):
-            input_output_map[kv_name_pair[0]] = kv_name_pair[1]
-            input = ov_model.input(kv_name_pair[0])
-            shape = input.get_partial_shape()
+    if enable_fuse_cache_reorder:
+        if not model_has_cache_reorder(ov_model):   # the transformation wasn't applied when model was exported
+            fuse_cache_reorder(ov_model, not_kv_inputs, hf_model.key_value_input_names, batch_dim)
+        else:
+            print('[ WARNING ] Model has "beam_idx" parameter which means that the cache reorder is already fused, skipping fuse transformation')
 
-            # By default, batch is the 0-th but chatglm uses 1-st dimension as batch
-            # TODO: Deduce from a model via ordinal reshape
-            batch_idx = 1 if hf_model.config.model_type == 'chatglm' else 0
-            shape[batch_idx] = batch_size * num_attention_heads * num_beams
+        if patch_methods:
+            # override _reorder_cache to avoid cache manipulation outside of the model as it is already done inside
+            def _reorder_cache_stub(self, past_key_values, beam_idx):
+                # TODO: Apply it differently based on model type
+                self.next_beam_idx = np.array(beam_idx)  # save beam_idx to be used as an input in the next iteration
+                return past_key_values
 
-            input.get_node().set_partial_shape(shape)
+            hf_model.use_cache_as_state = False
+            hf_model._reorder_cache = types.MethodType(_reorder_cache_stub, hf_model)
+            hf_model.forward = types.MethodType(forward_simplified, hf_model)  # need custom forward to set beam_idx input to OV model
+            hf_model.next_beam_idx = np.zeros([num_beams * batch_size], dtype=int)  # initial value for beam_idx is all zeros
```

Comment:
this default value has flaw,  for example when num_beams=4 and batch_size=2, next_beam_idx should be `[0,0,0,0,1,1,1,1]`  instead of all zeros